security/security.c

```diff
@@ -307,6 +382,50 @@ static void __init ordered_lsm_parse(const char *order, const char *origin)
 	kfree(sep);
 }
 
+static void __init lsm_init_hook_static_slot(struct security_static_slot *slots,
+					     struct hlist_head *head,
+					     int *first_slot_idx)
+{
+	struct security_hook_list *pos;
+	struct security_static_slot *slot;
+	int slot_cnt, first_slot;
+
+	slot_cnt = 0;
+	hlist_for_each_entry_rcu (pos, head, list) {
+		slot_cnt++;
+	}
+
+	first_slot = SECURITY_STATIC_SLOT_COUNT - slot_cnt;
+	if (first_slot < 0)
+		panic("%s - No static hook slot remaining to add LSM hook.\n",
+		      __func__);
+
+	if (slot_cnt == 0) {
+		*first_slot_idx = INT_MAX;
+		return;
+	}
+
+	slot = slots + first_slot;
+	hlist_for_each_entry_rcu (pos, head, list) {
+		__static_call_update(slot->key, slot->tramp,
+				     pos->hook.generic_func);
+		slot++;
+	}
+
+	// write only after the static calls are filled
```

Comment:
This comment sounds as if it's saying we do it in this order to avoid issues with concurrent access (i.e. to ensure that nobody reads `*first_slot_index` and then uses that to index a slot that hasn't been populated by this code yet). However writing the C code in this order will _not_ fix that issue, if it exists. The compiler and the CPU can both re-order the memory accesses so `*first_slot_index` effectively gets written before the slots are really populated. 

After a quick search the best resources I found about this were [here](https://github.com/google/ktsan/wiki/READ_ONCE-and-WRITE_ONCE#it-is-required-by-relevant-standards) and (warning - this one is very intense and covers much more ground than this specific case, but it's worth reading) [here](https://lwn.net/Articles/793253/). We can talk more about this in person too. This isn't just a C thing BTW - you'd run into comparable issues in Go and Java too.

Luckily, since this is all happening during init, I don't think it matters! You can just remove the comment.

Reply:
Good point, thanks!
It boils down to: can `security_<some_hook>` be called during the init? If yes, then we need to insure that the write to `*first_slot_index` happens after the slots are filled. Otherwise, `security_<some_hook>` could read the first_slot_index`, jump to the slot, and start calling the functions which are not there yet.